### PR TITLE
Feature/input tool

### PR DESCRIPTION
**Pull request by brandonzhu09**
## Description

Developed the initial version of the input tool module with preliminary testing that inserts plot components in the evagram database for Swell workflow tasks.

### CODE REVIEW

**Comment by mathomp4** [comment]
Ahh. Interesting. A couple things for @asewnath to ponder over. 

Because I started from the gmao template repo, this repo got both our label enforcer and our changelog enforcer. 

First: labels. That means it's looking for our usual "0-diff" etc. labels. That might not mean anything for evagram. So I could just remove the label enforcer. Or I could change it if there are any labels you'd like to require.

Second: changelog. The enforcer is trying to make sure every PR changes `CHANGELOG.md`. This is sort of nice so that when it comes time to release, etc., you have a changelog from the devs already there. We usually follow [Keep-A-Changelog format](https://keepachangelog.com/en/1.1.0/). Again, I could remove the enforcer or you could choose to use a changelog like this.

Third: I can make any of these required for PRs to merge. From what I see, you'd probably at least want Python Coding Norms and Yaml Lint required. And maybe changelog if you choose that.

**Comment by asewnath** [comment]
Hi @mathomp4 we can remove the label enforcer for evagram-input. As for the changelog, sure let's keep it.

For the required tests before PR merging, let's keep the coding norms, yaml linter, and change log tests.

@brandonzhu09 you can look at the documentation Matt pointed to or other repositories in GEOS-ESM to get an idea of how to update the changelog

**Comment by mathomp4** [comment]
Never mind. I took care of removing the label enforcer file in #3  :)

**Comment by brandonzhu09** [comment]
Thanks for all the information! I will update the Changelog to these standards before the PR gets merged.

**Comment by mathomp4** [comment]
@brandonzhu09 Okay. I merged in #3 which also removed the label enforcer. So once you press the "Update branch" button here you should get that in and then it won't look for them.

File: src/evagram/database/input_tool.py
Comment (asewnath) [inline src/evagram/database/input_tool.py]: Is there a reason to print a message and return 1 instead of using raising an exception?

File: src/evagram/database/input_tool.py
Comment (asewnath) [inline src/evagram/database/input_tool.py]: raise exception?

File: src/evagram/database/input_tool.py
Comment (asewnath) [inline src/evagram/database/input_tool.py]: Don't need to return 0

File: src/evagram/database/input_tool.py
Comment (asewnath) [inline src/evagram/database/input_tool.py]: Great job! Good to see the add plot function! Again, I don't think you need to return 0 here

File: src/evagram/database/input_tool.py
Comment (asewnath) [inline src/evagram/database/input_tool.py]: Why not fetch the user ID after inserting the table record instead of doing another function call? Is it optimal to do the recursion instead?

File: src/evagram/database/input_tool.py
Comment (asewnath) [inline src/evagram/database/input_tool.py]: same as `add_current_user`

File: tests/test_input_tool.py
Comment (asewnath) [inline tests/test_input_tool.py]: It's unclear whether we should keep all these tests in evagram_input. We'll have to think about this more. The purpose of this repository was to be a lightweight dependency for swell, it would be nice to have it not keep around large pkl files

File: src/evagram/database/input_tool.py
Comment (asewnath) [inline src/evagram/database/input_tool.py]: Can you change the directory structure to something like `src/evagram_input/input_tool.py`. Because input_tool is a script, I'm not sure we need another directory.

File: CHANGELOG.md
Comment (asewnath) [inline CHANGELOG.md:9]: `## [1.0.0] - 2024-03-29 ` above `### Added` to signify the first version of this input tool. You can remove all the other headers (i.e. `### Removed`) you don't have content for

**Comment by asewnath** [comment]
Not the most pressing concern, but can you also update the readme to say evagram_input?

File: tests/test_input_tool.py
Comment (ghost) [inline tests/test_input_tool.py]: Should an input tool have any `delete` tests?

File: tests/test_input_tool.py
Comment (ghost) [inline tests/test_input_tool.py]: Should an input tool have any `fetch` tests?

File: src/evagram/database/input_tool.py
Comment (ghost) [inline src/evagram/database/input_tool.py]: I think the Session class should be responsible for forming the connection.  See comment below about the cursor.

File: src/evagram/database/input_tool.py
Comment (ghost) [inline src/evagram/database/input_tool.py]: A couple notes on code styling:  Can you add some more comments here describing what the Session class is for and what it does.  You can refer to pep257 conventions, though we don't necessarily need to enforce that to a T.  Also, some spacing between lines or groups of lines will help with readability.

File: src/evagram/database/input_tool.py
Comment (ghost) [inline src/evagram/database/input_tool.py]: Is there a way we can rewrite this without so many indentations?

File: src/evagram/database/input_tool.py
Comment (ghost) [inline src/evagram/database/input_tool.py]: `open_session` is the driving force of this whole input operation.  I don't think it should be tucked into Session's init.  It should be called explicitly by the user.

Comment (ghost) [inline src/evagram/database/input_tool.py]: Additionally, is "open_session" the correct terminology?  It does a lot more than opening.  It does everything.

File: src/evagram/database/input_tool.py
Comment (ghost) [inline src/evagram/database/input_tool.py]: I don't feel comfortable with the cursor being tossed around to functions external to the Session class.  It should stay in house as a private attribute.  The Session class can have a private wrapper methods for `fetch_one`, `fetch_many`, `insert_table_record`, `form_connection`, `close_connection`, etc.  That's how I imagine a Session object's responsibilities.  It manages the most sensitive, sql-related details internally.

Comment (ghost) [inline src/evagram/database/input_tool.py]: Instead of standalone functions, can `insert` and `add` functions be private methods of the Session class?

File: tests/test_input_tool.py
Comment (ghost) [inline tests/test_input_tool.py]: Should an input tool have `join` tests?  Maybe this one can stay.

File: src/evagram/database/input_tool.py
Comment (ghost) [inline src/evagram/database/input_tool.py]: OR, another option is for the `insert` and `add` functions to remain external but to accept a Session object instead of the cursor.  My  main point is, the cursor should be protected and  unexposed.

**Review by ghost — COMMENTED** [review]
First round of comments.  We can talk more about this tomorrow.

File: src/evagram/database/input_tool.py
Comment (asewnath) [reply src/evagram/database/input_tool.py]: Let's definitely talk about this one tomorrow. I had suggested passing the connection into session and I think this suggestion was a relic from when we were planning on calling input_tool for the evagram database tests as well

**Comment by asewnath** [comment]
Add setup.py/ __init__.py files to make it into a module/ make github action :)

File: src/evagram/database/input_tool.py
Comment (asewnath) [reply src/evagram/database/input_tool.py]: open connection in session instead of passing it in

File: CHANGELOG.md
Comment (mathomp4) [inline CHANGELOG.md:13]: You can keep these lines in here. They are just "section placeholders" so that if someone Adds something or Changes something, they can put it under the right section.

Comment (asewnath) [reply CHANGELOG.md:13]: Thanks Matt, good to know!

File: CHANGELOG.md
Comment (asewnath) [reply CHANGELOG.md:9]: @brandonzhu09 Matt suggested we leave those headers in for future changes

**Comment by asewnath** [comment]
Some notes from the evagram meeting:
- see if a try/catch approach for checking if record exists
- update the readme to describe usage
- close cursor in input_data
- update tests to prove database rollback works
- create github action to test evagram_input
- wrapper function for Session

Here are some notes Tariq had provided regarding the wrapper function:

api.py:
```
from input_tool import input_tool
 
def input_data(**kwargs):
    session = input_tool.Session(**kwargs)
    session.input_data()
```

**Comment by ghost** [comment]
For a follow-up PR, consider returning a summary statement to the user after successfully completing `input_data()`

**Review by asewnath — DISMISSED** [review]

File: src/evagram_input/input_tool/input_tool.py
Comment (ghost) [inline src/evagram_input/input_tool/input_tool.py]: make private

File: src/evagram_input/input_tool/input_tool.py
Comment (ghost) [inline src/evagram_input/input_tool/input_tool.py]: make private

File: .env
Comment (ghost) [inline .env]: remove file

**Comment by brandonzhu09** [comment]
I did a little thinking on the central endpoint name and thought since the purpose of the module is to conduct an experiment in a swell workflow, we can instead call it `conduct_experiment` or something along the lines of that. What do we think?

**Comment by ghost** [comment]
> I did a little thinking on the central endpoint name and thought since the purpose of the module is to conduct an experiment in a swell workflow, we can instead call it `conduct_experiment` or something along the lines of that. What do we think?

Let's copy over this comment to issue #4

File: src/evagram_input/input_tool/input_tool.py
Comment (ghost) [reply src/evagram_input/input_tool/input_tool.py]: conn and cursor

File: src/evagram_input/input_tool/input_tool.py
Comment (ghost) [inline src/evagram_input/input_tool/input_tool.py]: print(err)

Comment (brandonzhu09) [reply src/evagram_input/input_tool/input_tool.py]: Fixed!

**Review by asewnath — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by mathomp4 — APPROVED** [review]

**Comment by brandonzhu09** [comment]
Thank you all for approving this PR! It says that I am not authorized to merge this PR so if I can get permission or someone can help me merge would be great.

**Pull request merged**
